Add legion host identifier

balag0 opened the pull request:
Subsequent PRs will customize different host services as required. Scope of this PR is to add a way to check if container is running on a Legion host.

<!-- Please provide all the information below.  -->

[### Issue describing the changes in this PR]
https://github.com/Azure/azure-functions-host/issues/8616

resolves #issue_for_this_pr

### Pull request checklist

* [X ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ X] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [X ] Otherwise: Backport tracked by issue/PR https://github.com/Azure/azure-functions-host/issues/8616
* [ X] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [ ] I have added all required tests (Unit tests, E2E tests)

mathewc commented on src/WebJobs.Script.WebHost/Program.cs:
```diff
@@ -87,7 +87,7 @@ public static IWebHostBuilder CreateWebHostBuilder(string[] args = null)
         private static void InitializeProcess()
         {
-            if (SystemEnvironment.Instance.IsLinuxConsumption())
+            if (SystemEnvironment.Instance.IsLinuxConsumptionOnAtlas() || SystemEnvironment.Instance.IsLinuxConsumptionOnLegion())
```
Can't you use IsAnyLinuxConsumption() here?

mathewc commented on src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs:
```diff
@@ -99,7 +99,11 @@ public static void AddWebJobsScriptHost(this IServiceCollection services, IConfi
             {
                 var environment = p.GetService<IEnvironment>();
-                if (environment.IsLinuxConsumption())
+                if (environment.IsLinuxConsumptionOnAtlas())
```
The logic below is the same for Atlas and Legion, so can't we just change the existing if clause to IsAnyLinuxConsumption()?

Tonewall replied:
Assuming this is a placeholder until we create the different eventgenerator for Legion?

balag0 replied:
Yes, we could use Any() here, I just wanted to highlight that some changes are needed here. 
basically by searching for IsLinuxConsumptionOnLegion we can identify the list of areas that need changes.
any code using Any() is already using the right logic and no further changes are required.
(same reasoning applies to the comment in InitializeProcess as well)

@Tonewall Correct. this PR just highlights the places needing changes.

mathewc replied:
Ok, then if you're saying some points in this code require further PRs to complete, then please add TODOs on the logic in question so it's clear.

mathewc commented on src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs:
```diff
@@ -214,7 +218,7 @@ private static void AddLinuxContainerServices(this IServiceCollection services)
             {
                 var environment = s.GetService<IEnvironment>();
-                if (environment.IsLinuxConsumption())
+                if (environment.IsLinuxConsumptionOnAtlas() || environment.IsLinuxConsumptionOnLegion())
```
IsAnyLinuxConsumption()?

balag0 replied:
same as https://github.com/Azure/azure-functions-host/pull/8617#discussion_r942991081
basically any code using Any will stay that way
any code using specific checks will need to change.

mathewc replied:
Then please put TODOs, since as it stands if this PR is checked in people will have the same question I did :)

mathewc commented on src/WebJobs.Script/Environment/EnvironmentExtensions.cs:
```diff
@@ -254,9 +254,22 @@ public static bool IsWindowsAzureManagedHosting(this IEnvironment environment)
+
+        public static bool IsLinuxConsumptionOnLegion(this IEnvironment environment)
+        {
+            return !string.IsNullOrEmpty(environment.GetEnvironmentVariable(WebsitePodName)) &&
```
Should this also have !environment.IsAppService()?

Tonewall replied:
Also a case if ContainerName exists?

balag0 replied:
I think LegionHost check should cover IsAppService check as well. basically they are mutually exclusive already

it is okay for a container to have both pod name and containername. the presence of legionhost appsetting is the deciding factor.

Tonewall commented on src/WebJobs.Script/Environment/EnvironmentExtensions.cs:
```diff
@@ -30,7 +30,7 @@ public static string GetEnvironmentVariableOrDefault(this IEnvironment environme
         public static bool IsLinuxMetricsPublishingEnabled(this IEnvironment environment)
         {
-            return environment.IsLinuxConsumption() && string.IsNullOrEmpty(environment.GetEnvironmentVariable(ContainerStartContext));
+            return environment.IsLinuxConsumptionOnAtlas() && string.IsNullOrEmpty(environment.GetEnvironmentVariable(ContainerStartContext));
```
Should this be IsAnyLinuxConsumption()?

balag0 replied:
please see https://github.com/Azure/azure-functions-host/pull/8617#discussion_r942991081

Tonewall replied:
Since we will implement a similar metric logging mechanism for Legion, shouldn't this line also include legion runtime? So maybe something like `environment.IsLinuxConsumptionOnAtlas() || environment.IsLinuxConsumptionOnLegion()` for right now?

balag0 replied:
yeah, I was thinking we will be modifying the existing service, so  didnt add a todo here explicitly, but we know this pending work anyway :)

mathewc commented on src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs:
```diff
@@ -214,7 +219,8 @@ private static void AddLinuxContainerServices(this IServiceCollection services)
             {
                 var environment = s.GetService<IEnvironment>();
-                if (environment.IsLinuxConsumption())
+                //todo: Replace with legion specific service
```
this comment may be misleading? When legion logic is added, you'll be adding a new if branch here, not replacing all this logic right?

Also - is there a tracking issue for doing the follow up work tracked by these TODOs? It'd be good to have that issue and reference it in all these TODOs so it's all linked up and can be cleaned up in one go.

balag0 replied:
Yes, will start adding some tasks here. currently they are tracked in dev ops.

balag0 committed:
Add legion host identifier

Subsequent PRs will customize different services as required. Scope of this PR is to add a way to check if container is running on a Legion host.

balag0 commented:
/backport to v3.x
